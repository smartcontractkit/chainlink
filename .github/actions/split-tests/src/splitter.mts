import {Tests, TestsBySplit} from "./types.mjs";

/**
 * Split tests by first prioritizing slow tests being spread over each split, then filling each split by test list order.
 *
 * @example
 * Given the following arguments:
 * tests: ['foo.test', 'bar.test', 'baz.test', 'yup.test', 'nope.test']
 * slowTests: ['bonk.test', 'bop.test', 'ouch.test.ts']
 * numOfSplits: 2
 *
 * We get the following output:
 * 1. Spread slow tests across splits: [['bonk.test', 'ouch.test.ts'], ['bop.test']]
 * 2. Insert list of tests: [['bonk.test', 'ouch.test.ts', 'foo.test', 'bar.test'], ['bop.test', 'baz.test', 'yup.test', 'nope.test']]
 *
 * @param tests A list of tests to distribute across splits by the test list order
 * @param slowTests A list of slow tests, where the list of tests is evenly distributed across all splits before inserting regular tests
 * @param numOfSplits The number of splits to spread tests across
 */
export function simpleSplit(
  tests: Tests,
  slowTests: Tests,
  numOfSplits: number
): TestsBySplit {
  const maxTestsPerSplit = Math.max(tests.length / numOfSplits);

  const testsBySplit: TestsBySplit = new Array(numOfSplits)
    .fill(null)
    .map(() => []);

  // Evenly distribute slow tests over each split
  slowTests.forEach((test, i) => {
    const splitIndex = i % numOfSplits;
    testsBySplit[splitIndex].push(test);
  });

  tests.forEach((test, i) => {
    const splitIndex = Math.floor(i / maxTestsPerSplit);
    testsBySplit[splitIndex].push(test);
  });

  return testsBySplit;
}
