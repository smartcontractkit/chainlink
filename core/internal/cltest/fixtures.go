package cltest

import (
	"encoding/json"
	"testing"

	"github.com/smartcontractkit/chainlink/core/store/models"

	"github.com/ethereum/go-ethereum/core/types"
	"github.com/stretchr/testify/require"
	"github.com/tidwall/gjson"
)

// JSONFromFixture create models.JSON from file path
func JSONFromFixture(t *testing.T, path string) models.JSON {
	return JSONFromBytes(t, MustReadFile(t, path))
}

// LogFromFixture create ethtypes.log from file path
func LogFromFixture(t *testing.T, path string) types.Log {
	value := gjson.Get(string(MustReadFile(t, path)), "params.result")
	var el types.Log
	require.NoError(t, json.Unmarshal([]byte(value.String()), &el))

	return el
}

// TxReceiptFromFixture create ethtypes.log from file path
func TxReceiptFromFixture(t *testing.T, path string) *types.Receipt {
	jsonStr := JSONFromFixture(t, path).Get("result").String()

	var receipt types.Receipt
	err := json.Unmarshal([]byte(jsonStr), &receipt)
	require.NoError(t, err)

	return &receipt
}
