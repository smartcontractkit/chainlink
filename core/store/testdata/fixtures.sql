INSERT INTO keys (address, json, created_at, updated_at, next_nonce) VALUES (
    E'\\x3cb8e3FD9d27e39a5e9e6852b0e96160061fd4ea',
    '{"address":"3cb8e3fd9d27e39a5e9e6852b0e96160061fd4ea","crypto":{"cipher":"aes-128-ctr","ciphertext":"7515678239ccbeeaaaf0b103f0fba46a979bf6b2a52260015f35b9eb5fed5c17","cipherparams":{"iv":"87e5a5db334305e1e4fb8b3538ceea12"},"kdf":"scrypt","kdfparams":{"dklen":32,"n":262144,"p":1,"r":8,"salt":"d89ac837b5dcdce5690af764762fe349d8162bb0086cea2bc3a4289c47853f96"},"mac":"57a7f4ada10d3d89644f541c91f89b5bde73e15e827ee40565e2d1f88bb0ac96"},"id":"c8cb9bc7-0a51-43bd-8348-8a67fd1ec52c","version":3}',
    '2019-01-01',
    '2019-01-01',
    0
);
INSERT INTO users (email, hashed_password, token_secret, created_at, updated_at) VALUES (
    'apiuser@chainlink.test',
    '$2a$10$bbwErtZcZ6qQvRsfBiY2POvuY6D4lwj/Vxq/PcVAL6o64nRaPgaEa', -- hash of literal string 'password'
    '1eCP/w0llVkchejFaoBpfIGaLRxZK54lTXBCT22YLW+pdzE4Fafy/XO5LoJ2uwHi',
    '2019-01-01',
    '2019-01-01'
);
