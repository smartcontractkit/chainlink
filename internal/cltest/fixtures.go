package cltest

const (
	key3cb8e3fd9d27e39a5e9e6852b0e96160061fd4ea = `{"address":"3cb8e3fd9d27e39a5e9e6852b0e96160061fd4ea","crypto":{"cipher":"aes-128-ctr","ciphertext":"7515678239ccbeeaaaf0b103f0fba46a979bf6b2a52260015f35b9eb5fed5c17","cipherparams":{"iv":"87e5a5db334305e1e4fb8b3538ceea12"},"kdf":"scrypt","kdfparams":{"dklen":32,"n":262144,"p":1,"r":8,"salt":"d89ac837b5dcdce5690af764762fe349d8162bb0086cea2bc3a4289c47853f96"},"mac":"57a7f4ada10d3d89644f541c91f89b5bde73e15e827ee40565e2d1f88bb0ac96"},"id":"c8cb9bc7-0a51-43bd-8348-8a67fd1ec52c","version":3}`
)
