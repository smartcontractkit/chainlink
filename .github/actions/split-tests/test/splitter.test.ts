import {simpleSplit} from "../src/splitter.mjs";
import {testArr, testSievedArr, testSlowArr} from "./fixtures.mjs";

describe("simpleSplit", () => {
  it("doesn't error on empty arrays", () => {
    expect(simpleSplit([], [], 1)).toMatchSnapshot();
    expect(simpleSplit([], [], 5)).toMatchSnapshot();
  });

  it("handles no slow test splitting", () => {
    expect(simpleSplit(testArr, [], 1)).toMatchSnapshot();
    expect(simpleSplit(testArr, [], 2)).toMatchSnapshot();
    expect(simpleSplit(testArr, [], 3)).toMatchSnapshot();
  });

  it("handles slow test splitting", () => {
    expect(simpleSplit(testSievedArr, testSlowArr, 1)).toMatchSnapshot();
    expect(simpleSplit(testSievedArr, testSlowArr, 2)).toMatchSnapshot();
    expect(simpleSplit(testSievedArr, testSlowArr, 3)).toMatchSnapshot();
  });
});
