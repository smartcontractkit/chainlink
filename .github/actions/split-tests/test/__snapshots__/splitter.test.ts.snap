// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simpleSplit doesn't error on empty arrays 1`] = `
[
  [],
]
`;

exports[`simpleSplit doesn't error on empty arrays 2`] = `
[
  [],
  [],
  [],
  [],
  [],
]
`;

exports[`simpleSplit handles no slow test splitting 1`] = `
[
  [
    "foo.test",
    "bar.test",
    "baz.test",
    "yup.test",
    "nope.test",
    "bonk.test",
    "bop.test",
    "ouch.test.ts",
  ],
]
`;

exports[`simpleSplit handles no slow test splitting 2`] = `
[
  [
    "foo.test",
    "bar.test",
    "baz.test",
    "yup.test",
  ],
  [
    "nope.test",
    "bonk.test",
    "bop.test",
    "ouch.test.ts",
  ],
]
`;

exports[`simpleSplit handles no slow test splitting 3`] = `
[
  [
    "foo.test",
    "bar.test",
    "baz.test",
  ],
  [
    "yup.test",
    "nope.test",
    "bonk.test",
  ],
  [
    "bop.test",
    "ouch.test.ts",
  ],
]
`;

exports[`simpleSplit handles slow test splitting 1`] = `
[
  [
    "bonk.test",
    "bop.test",
    "ouch.test.ts",
    "foo.test",
    "bar.test",
    "baz.test",
    "yup.test",
    "nope.test",
  ],
]
`;

exports[`simpleSplit handles slow test splitting 2`] = `
[
  [
    "bonk.test",
    "ouch.test.ts",
    "foo.test",
    "bar.test",
    "baz.test",
  ],
  [
    "bop.test",
    "yup.test",
    "nope.test",
  ],
]
`;

exports[`simpleSplit handles slow test splitting 3`] = `
[
  [
    "bonk.test",
    "foo.test",
    "bar.test",
  ],
  [
    "bop.test",
    "baz.test",
    "yup.test",
  ],
  [
    "ouch.test.ts",
    "nope.test",
  ],
]
`;
